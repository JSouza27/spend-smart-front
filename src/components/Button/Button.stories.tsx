import { ComponentMeta, ComponentStory } from '@storybook/react';
import { ThemeProvider } from 'styled-components';
import { Plus } from '@styled-icons/bootstrap';

import Button from '.';
import theme from '../../styles/theme';
import { CSSProperties } from 'react';

export default {
  title: 'Button',
  component: Button,
  args: {
    children: 'Nova Transação',
    size: 'medium',
    appearance: 'primary'
  }
} as ComponentMeta<typeof Button>;

const styles: CSSProperties = {
  alignItems: 'center',
  display: 'flex',
  justifyContent: 'center',
  height: '90vh',
  width: '100%'
};

export const Basic: ComponentStory<typeof Button> = (args) => (
  <ThemeProvider theme={theme}>
    <div style={styles}>
      <Button {...args} />
    </div>
  </ThemeProvider>
);

export const WithIcon: ComponentStory<typeof Button> = (args) => (
  <ThemeProvider theme={theme}>
    <div style={styles}>
      <Button icon={<Plus />} {...args} />
    </div>
  </ThemeProvider>
);

export const Appearance: ComponentStory<typeof Button> = (args) => (
  <ThemeProvider theme={theme}>
    <div style={styles}>
      <Button {...args} />
    </div>
  </ThemeProvider>
);
