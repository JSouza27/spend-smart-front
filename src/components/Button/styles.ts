import styled, { DefaultTheme, css } from 'styled-components';
import { ButtonProps } from '.';

type WrapperProps = {
  hasIcon: boolean;
} & Pick<ButtonProps, 'size' | 'appearance'>;

const wrapperModifier = {
  small: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.xsmall};
    height: 2.4rem;
    padding: ${theme.spacings.xxxsmall} 1.2rem;
    line-height: ${theme.spacings.small};
  `,
  medium: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.xsmall};
    height: 3.2rem;
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xsmall};
    line-height: ${theme.spacings.medium};
  `,
  large: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.small};
    height: 4rem;
    padding: 1.2rem ${theme.spacings.small};
    line-height: ${theme.spacings.large};
  `,
  withIcon: () => css`
    svg {
      color: inherit;
      width: 2rem;
    }
  `
};

const wrapperModifierColor = {
  primary: (theme: DefaultTheme) => css`
    background-color: ${theme.colors.blue_400};
    color: ${theme.colors.neutral_0};
    outline-color: ${theme.colors.blue_200};

    &:hover {
      background-color: ${theme.colors.blue_500};
    }

    &:active {
      background-color: ${theme.colors.blue_600};
    }

    &:focus {
      background-color: ${theme.colors.blue_500};
      border: 2px solid ${theme.colors.blue_200};
    }

    &:disabled {
      background-color: ${theme.colors.blue_500};
    }
  `,
  secondary: (theme: DefaultTheme) => css`
    background-color: ${theme.colors.neutral_0};
    border: 1px solid ${theme.colors.neutral_400};
    color: ${theme.colors.neutral_700};
    outline-color: ${theme.colors.blue_200};

    &:hover {
      border-color: ${theme.colors.neutral_600};
      color: ${theme.colors.neutral_800};
    }

    &:active {
      background-color: ${theme.colors.neutral_100};
      border-color: ${theme.colors.neutral_500};
      color: ${theme.colors.neutral_800};
    }

    &:focus {
      background-color: ${theme.colors.neutral_0};
      border-width: 2px;
      color: ${theme.colors.neutral_800};
    }

    &:disabled {
      border-color: ${theme.colors.neutral_300};
      color: ${theme.colors.neutral_500};
    }
  `,
  minimal: (theme: DefaultTheme) => css`
    background-color: transparent;
    color: ${theme.colors.neutral_700};
    outline-color: ${theme.colors.blue_200};

    &:hover {
      background-color: ${theme.colors.neutral_100};
      color: ${theme.colors.neutral_800};
    }

    &:active {
      background-color: ${theme.colors.neutral_200};
      color: ${theme.colors.neutral_800};
    }

    &:focus {
      background-color: ${theme.colors.neutral_0};
      color: ${theme.colors.neutral_800};
    }

    &:disabled {
      color: ${theme.colors.neutral_400};
    }
  `,
  error: (theme: DefaultTheme) => css`
    background-color: ${theme.colors.red_400};
    color: ${theme.colors.neutral_0};
    outline-color: ${theme.colors.red_100};

    &:hover {
      background-color: ${theme.colors.red_500};
    }

    &:active {
      background-color: ${theme.colors.red_600};
    }

    &:focus {
      background-color: ${theme.colors.red_500};
    }

    &:disabled {
      background-color: ${theme.colors.red_100};
    }
  `
};

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, size, hasIcon, appearance }) => css`
    align-items: center;
    border: 0;
    border-radius: ${theme.border.radius};
    cursor: pointer;
    display: flex;
    font-weight: ${theme.font.semi_bold};
    gap: ${theme.spacings.xxsmall};
    white-space: nowrap;

    ${!!size && wrapperModifier[size](theme)};
    ${!!hasIcon && wrapperModifier.withIcon};
    ${!!appearance && wrapperModifierColor[appearance](theme)}
  `}
`;
