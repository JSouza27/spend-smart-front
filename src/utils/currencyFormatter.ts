export class CurrencyFormatter {
  private static language = 'pt-BR';
  private static currency = 'BRL';

  static formatter(num: number): string {
    return (num ?? 0).toLocaleString(this.language, {
      currency: this.currency
    });
  }

  static unformat(num: string): number {
    const nums = num.replace(/[^0-9]+/g, '');
    const i = nums.length - 2;

    return Number(`${nums.substring(0, i)}.${nums.substring(i)}`);
  }
}
