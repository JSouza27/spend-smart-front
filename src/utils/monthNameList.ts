export const monthNameList = [
  'janeiro',
  'fevereiro',
  'mar√ßo',
  'abril',
  'maio',
  'junho',
  'julho',
  'agosto',
  'setembro',
  'outubro',
  'novembro',
  'dezembro'
];
