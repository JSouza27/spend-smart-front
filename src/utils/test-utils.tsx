import { render } from '@testing-library/react';
import React from 'react';
import { ThemeProvider } from 'styled-components';

import theme from '../styles/theme';
import { FormProvider, useForm } from 'react-hook-form';

const AllTheProviders = ({ children }: any) => {
  const methods = useForm();

  return (
    <ThemeProvider theme={theme}>
      <FormProvider {...methods}>{children}</FormProvider>
    </ThemeProvider>
  );
};

const customRender = (
  ui: React.ReactElement<any, string | React.JSXElementConstructor<any>>,
  options?: any
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
