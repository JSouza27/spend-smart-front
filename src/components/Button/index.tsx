import React, { ButtonHTMLAttributes, HtmlHTMLAttributes } from 'react';
import * as S from './styles';

export type ButtonProps = {
  children?: React.ReactNode;
  size?: 'small' | 'medium' | 'large';
  icon?: React.ReactNode;
  appearance?: 'primary' | 'secondary' | 'minimal' | 'error';
} & ButtonHTMLAttributes<HTMLButtonElement>;

export default function Button({
  children,
  size = 'medium',
  icon,
  appearance = 'primary',
  ...props
}: ButtonProps) {
  return (
    <S.Wrapper size={size} hasIcon={!!icon} {...props} appearance={appearance}>
      {!!icon && icon}
      {!!children && <span>{children}</span>}
    </S.Wrapper>
  );
}
