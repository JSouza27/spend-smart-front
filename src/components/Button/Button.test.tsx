import { Plus } from '@styled-icons/bootstrap';

import { render, screen } from 'utils/test-utils';
import Button from '.';

describe('<Button />', () => {
  it('should render the medium size by default', () => {
    const { container } = render(<Button>Nova Transição</Button>);

    const button = screen.getByRole('button', { name: /nova transição/i });

    expect(button).toBeInTheDocument();
    expect(button).toHaveStyle({
      height: '3.2rem',
      'font-size': '1.2rem',
      padding: '0.8rem 1.6rem'
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the small size', () => {
    render(<Button size="small">Nova Transição</Button>);

    const button = screen.getByRole('button', { name: /nova transição/i });

    expect(button).toBeInTheDocument();
    expect(button).toHaveStyle({
      height: '2.4rem',
      'font-size': '1.2rem',
      padding: '0.4rem 1.2rem'
    });
  });

  it('should render the large size', () => {
    render(<Button size="large">Nova Transição</Button>);

    const button = screen.getByRole('button', { name: /nova transição/i });

    expect(button).toBeInTheDocument();
    expect(button).toHaveStyle({
      height: '4rem',
      'font-size': '1.4rem',
      padding: '1.2rem 2.4rem'
    });
  });

  it('should render an icon version', () => {
    render(
      <Button icon={<Plus data-testid="icon" />} size="medium">
        Nova Transição
      </Button>
    );

    const button = screen.getByText(/nova transição/i);
    const icon = screen.getByTestId(/icon/i);

    expect(button).toBeInTheDocument();
    expect(icon).toBeInTheDocument();
  });

  it('should render an icon version', () => {
    render(
      <Button icon={<Plus data-testid="icon" />} size="medium">
        Nova Transição
      </Button>
    );

    const button = screen.getByText(/nova transição/i);
    const icon = screen.getByTestId(/icon/i);

    expect(button).toBeInTheDocument();
    expect(icon).toBeInTheDocument();
  });
});
