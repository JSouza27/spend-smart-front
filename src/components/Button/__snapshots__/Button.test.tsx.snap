// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the medium size by default 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 0;
  border-radius: 0.4rem;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-weight: 600;
  gap: 0.8rem;
  white-space: nowrap;
  font-size: 1.2rem;
  height: 3.2rem;
  padding: 0.8rem 1.6rem;
  line-height: 3.2rem;
  background-color: #3366FF;
  color: #FFFFFF;
  outline-color: #D6E0FF;
}

.c0:hover {
  background-color: #2952CC;
}

.c0:active {
  background-color: #1F3D99;
}

.c0:focus {
  background-color: #2952CC;
  border: 2px solid #D6E0FF;
}

.c0:disabled {
  background-color: #2952CC;
}

<button
  class="c0"
>
  <span>
    Nova Transição
  </span>
</button>
`;
